api/src/main/java/edu/wpi/first/shuffleboard/api/util/ThreadUtils.java

```diff
@@ -27,4 +29,36 @@ public static Thread makeDaemonThread(Runnable runnable) {
     return thread;
   }
 
+  /**
+   * Runs a function with a synchronization lock.
+   *
+   * @param lock     the synchronization lock to use
+   * @param function the function to run
+   */
+  public static void withLock(Lock lock, Runnable function) {
+    try {
+      lock.lock();
+      function.run();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Runs a function with a synchronization lock and returns its result.
+   *
+   * @param lock     the synchronization lock to use
+   * @param function the function to run
+   * @param <T>      the type of the value returned by the function
+   *
+   * @return the output of {@code function}
+   */
+  public static <T> T withLock(Lock lock, Supplier<? extends T> function) {
+    try {
+      lock.lock();
+      return function.get();
+    } finally {
+      lock.unlock();
+    }
+  }
```

Comment:
You could also create a decorator over `Lock` that adds these methods to the interfaces as well. Just a thought.

This is the interface:
```java
public interface ReadWriteLock {
    Lock readLock();

    Lock writeLock();
}
```

Just create your own implementation:

```java
public interface FunctionalReadWriteLock extends ReaWriteLock {
   @Override
   FunctionalLock readLock();
   @Override
   FunctionalLock writeLock();
}
```
Ect...
Or you could do something like this:
https://github.com/debezium/debezium/blob/master/debezium-core/src/main/java/io/debezium/util/FunctionalReadWriteLock.java